bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/regexp/CharReader.java

```diff
@@ -0,0 +1,103 @@
+/*
+ *  Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package io.ballerina.runtime.internal.regexp;
+
+import io.ballerina.tools.text.TextDocument;
+
+import java.util.Arrays;
+
+/**
+ * A character reader utility used by TreeTraverser when parsing Ballerina string value of a
+ * regular expression string.
+ *
+ * @since 2201.3.0
+ */
+public class CharReader {
```

Comment:
@lasinicl can you point the issue that's tracking the refactoring for both parsers?

Reply:
https://github.com/ballerina-platform/ballerina-lang/issues/37735